### PR TITLE
feat(Context Based Restrictions): add service to project

### PR DESCRIPTION
**Pull request by JonahFarc**
## PR summary
Added Context-Based Restrictions API

## PR Checklist
Please make sure that your PR fulfills the following requirements:  
- [X] The commit message follows the [Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## Current vs new behavior  
Added Context-Based Restrictions API

## Does this PR introduce a breaking change?    
- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by Andras-Csanyi** [comment]
@JonahFarc the build has failed with checkstyle issues. Could you take a look at it and fix it?
I also like to ask a screenshot about green integration test results. Thank you!

File: modules/context-based-restrictions/src/test/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsIT.java [hunk 1]
```diff
@@ -0,0 +1,805 @@
...
+                    .build();
+
+            // Invoke operation
+            Response<OutZone> response = service.getZone(getZoneOptions).execute();
+        } catch (ServiceResponseException e) {
+            if (e.getStatusCode() != 404)
+                fail(String.format("Service returned status code %d: %s%nError details: %s",
+                        e.getStatusCode(), e.getMessage(), e.getDebuggingInfo()));
+        }
+    }
```
Comment (Andras-Csanyi) [inline modules/context-based-restrictions/src/test/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsIT.java:267]: I'd like to ask a test which exercises status code `400`. Go and Node sdks can be a good example.

File: modules/context-based-restrictions/src/test/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsIT.java [hunk 2]
```diff
@@ -0,0 +1,805 @@
...
+
+            // Invoke operation
+            Response<OutZone> response = service.replaceZone(replaceZoneOptions).execute();
+        } catch (ServiceResponseException e) {
+            if (e.getStatusCode() != 404)
+                fail(String.format("Service returned status code %d: %s%nError details: %s",
+                        e.getStatusCode(), e.getMessage(), e.getDebuggingInfo()));
+        }
+    }
+
```
Comment (Andras-Csanyi) [inline modules/context-based-restrictions/src/test/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsIT.java:329]: The Go and Node SDK contains tests covering `400` and `412`. I'd like to ask you to add tests here covering these status codes.

File: modules/context-based-restrictions/src/test/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsIT.java (hunk 1 above)
Comment (JonahFarc) [reply modules/context-based-restrictions/src/test/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsIT.java:268]: The Go and Node examples use a Missing Parameter error to catch this, unfortunately this does not work in java as it fails in the builder, before we can even make the request

File: modules/context-based-restrictions/src/test/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsIT.java (hunk 2 above)
Comment (JonahFarc) [reply modules/context-based-restrictions/src/test/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsIT.java:330]: Same as above for the 400, but the 412 I have added

File: modules/context-based-restrictions/pom.xml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        <groupId>com.ibm.cloud</groupId>
+        <version>99-SNAPSHOT</version>
+        <relativePath>../..</relativePath>
+    </parent>
+
+    <!-- >>> Replace this with the service module's artifactId (e.g. "example-service") -->
+    <artifactId>context-based-restrictions</artifactId>
+
+    <!-- >>> Replace this with a text description of this module (e.g. "Example Service") -->
+    <name>Context-Based Restrictions</name>
```
Comment (padamstx) [inline modules/context-based-restrictions/pom.xml:17]:
```suggestion
    <name>IBM Cloud Context-Based Restrictions API</name>
```
(just to be consistent with the other modules)

**Review by padamstx — APPROVED** [review]
LGTM, just one little tweak to the new service's pom.xml to make it consistent with the other modules.  I'll approve now to avoid a re-review.

File: modules/examples/src/main/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsExamples.java [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+      System.out.printf("deleteZone() response status code: %d%n", response.getStatusCode());
+    } catch (ServiceResponseException e) {
+        logger.error(String.format("Service returned status code %s: %s%nError details: %s",
+          e.getStatusCode(), e.getMessage(), e.getDebuggingInfo()), e);
+    }
+
+    try {
+      // begin-delete_rule
+      DeleteRuleOptions deleteRuleOptions = new DeleteRuleOptions.Builder()
+        .ruleId("testString")
```
Comment (padamstx) [inline modules/examples/src/main/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsExamples.java:309]: need to use a real rule id here

**Review by padamstx — CHANGES_REQUESTED** [review]
I suggested changes in the examples code to enable all the examples to run successfully

File: modules/examples/src/main/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsExamples.java [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+    @SuppressWarnings("checkstyle:methodlength")
+    public static void main(String[] args) throws Exception {
+        ContextBasedRestrictions contextBasedRestrictionsService = ContextBasedRestrictions.newInstance();
+        String serviceURL = "";
+        String testAccountID = "";
+        String testServiceName = "";
+        String zoneID = "";
+        String zoneRev = "";
+        String ruleID = "";
+        String ruleRev = "";
```
Comment (Andras-Csanyi) [inline modules/examples/src/main/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsExamples.java:79]:
```suggestion
        String serviceURL = null;
        String testAccountID = null;
        String testServiceName = null;
        String zoneID = null;
        String zoneRev = null;
        String ruleID = null;
        String ruleRev = null;
```

File: modules/examples/src/main/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsExamples.java [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+                    .name("SDK TEST - an example of zone")
+                    .accountId(testAccountID)
+                    .description("SDK TEST - this is an example of zone")
+                    .addresses(new java.util.ArrayList<Address>(java.util.Arrays.asList(addressModel)))
+                    .build();
+
+            Response<Zone> response = contextBasedRestrictionsService.createZone(createZoneOptions).execute();
+            Zone zone = response.getResult();
+
+            System.out.println(zone);
```
Comment (Andras-Csanyi) [inline modules/examples/src/main/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsExamples.java:104]:
```suggestion
            System.out.println(zone);
            // end-create_zone
```

Comment (Andras-Csanyi) [reply modules/examples/src/main/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsExamples.java:104]: This way we ensure that the 105 and 106 lines will not be displayed in the reference, so our users won't be confused what these lines for.

File: modules/examples/src/main/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsExamples.java [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+        } catch (ServiceResponseException e) {
+            logger.error(String.format("Service returned status code %s: %s%nError details: %s",
+                    e.getStatusCode(), e.getMessage(), e.getDebuggingInfo()), e);
+        }
+
+        try {
+            System.out.println("listZones() result:");
+            // begin-list_zones
+            ListZonesOptions listZonesOptions = new ListZonesOptions.Builder()
+                    .accountId(testAccountID)
```
Comment (Andras-Csanyi) [inline modules/examples/src/main/java/com/ibm/cloud/platform_services/context_based_restrictions/v1/ContextBasedRestrictionsExamples.java:117]: I'd like to ask you to rename this variable to `accountId` if it doesn't conflict. The point is not displaying `test` in the reference.

**Review by Andras-Csanyi — CHANGES_REQUESTED** [review]
Overall looks very good! I added a few comments too. Once these are fixed along with @padamstx comments we will be in a very good position to merge this PR in.

Thank you very much for the effort you put into it. Especially, the negative test cases.
Here comes the point when I have to apologise for being very picky in this review. :)

**Review by Andras-Csanyi — APPROVED** [review]
LGTM

**Review by padamstx — APPROVED** [review]
Overall looks good, although I suggested a few tweaks...
I'll approve now to avoid a re-review assuming those are addressed

**Pull request merged**